engines/docker/resultset.go

```diff
@@ -31,115 +41,189 @@ func (r *resultSet) Success() bool {
+		if result != nil {
+			result.Close()
+		}
+		return nil, err
 	}
-
-	_, err = tempfile.Seek(0, 0)
-	if err != nil {
-		monitor.ReportError(err, "could not seek to start of file")
-		return nil, runtime.ErrNonFatalInternalError
+	// If we had no error, and didn't get any response, we assume it's not found
+	if result == nil {
+		return nil, engines.ErrResourceNotFound
 	}
-
-	return tempfile, nil
+	return result, nil
 }
 
 func (r *resultSet) ExtractFolder(path string, handler engines.FileHandler) error {
-	path = filepath.Clean(path)
-	monitor := r.monitor.WithTag("extract-folder", path)
-
-	tarfile, err := r.extractFromContainer(path)
-	if err != nil {
-		return engines.ErrResourceNotFound
+	// We'll treat paths ending with a slash as paths to folders
+	if !strings.HasSuffix(path, "/") {
+		debug("ExtractFolder(%s) doesn't end with '/' which can't be a folder", path)
+		return runtime.NewMalformedPayloadError(fmt.Sprintf(
+			"docker folder path: '%s' doesn't end with slash, paths to folders must end with a slash", path,
+		))
 	}
-
-	defer func() {
-		err = tarfile.Close()
-		if err != nil {
-			monitor.ReportWarning(err, "could not close temporary file")
+	return r.extractResource(path, func(name string, stream ioext.ReadSeekCloser) error {
+		// Make the name relative to path
+		if !strings.HasPrefix(name, "/") {
+			name = "/" + name // Ensure we always have an absolute path
 		}
-	}()
+		name = name[len(path):]
+		return handler(name, stream)
+	})
+}
 
-	strip := filepath.Base(path) + "/"
-	_, err = tarfile.Seek(0, 0)
-	if err != nil {
-		monitor.ReportError(err, "could not seek to start of tar file")
-		return runtime.ErrNonFatalInternalError
+func (r *resultSet) extractResource(path string, handler engines.FileHandler) error {
+	// We force path to be absolute, this is the only sane thing
+	if !strings.HasPrefix(path, "/") {
+		debug("extractResource(%s) doesn't start with '/', hence it is a relative path", path)
+		return runtime.NewMalformedPayloadError(fmt.Sprintf(
+			"docker path: '%s' is a relative path, paths must be absolute", path,
+		))
 	}
-	reader := tar.NewReader(tarfile)
-	// Instead of using runtime.Untar it seems simpler
-	// to unpack each file one at a time and pass it to
-	// the handler.
-	for {
-		hdr, err := reader.Next()
-		if err != nil {
-			break
-		}
-		if hdr.Typeflag == tar.TypeDir {
-			continue
-		}
-		fname := strings.TrimPrefix(hdr.Name, strip)
-		m := monitor.WithTag("filename", fname)
 
-		tempfile, err := r.storage.NewFile()
-		if err != nil {
-			m.ReportError(err, "could not create temporary file")
-			continue
-		}
+	// Read tar stream while we download from docker
+	stream, streamWriter := io.Pipe()
 
-		if _, err = io.Copy(tempfile, reader); err != nil {
-			m.ReportError(err, "could not copy file")
-			continue
-		}
+	ctx, cancel := context.WithCancel(r.context)
+	notFound := false
+	var derr, rerr error
+	var interrupted atomics.Once
+	util.Parallel(func() {
+		defer cancel()
```

Comment:
Why do we need to unconditionally call `cancel`?

Reply:
Yes, contexts should **always** be canceled.

In this case, the code-path that handles reading the tar-stream should cancel the context given when fetching the tar-stream from docker. Because if the reading of the tar-stream fails, we want to abort the `DownloadFromDocker` call.

Reply:
I mean `DownloadFromContainer` should be aborted. Usually, it'll always be done, before we're doing reading the tar-stream... but just in case there is an error reading the tar-stream :)

Reply:
> Why do we need to unconditionally call cancel?

Oh, yeah, you are right.